test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -710,3 +790,217 @@ func TestBlockingOwnerRefDoesBlock(t *testing.T) {
 		t.Errorf("expect there to be 1 pods, got %#v", pods.Items)
 	}
 }
+
+// TestCustomResourceCascadingDeletion ensures the basic cascading delete
+// behavior supports custom resources.
+func TestCustomResourceCascadingDeletion(t *testing.T) {
+	ctx := setup(t, 5)
+	defer ctx.tearDown()
+
+	clientSet, apiExtensionClient, clientPool := ctx.clientSet, ctx.apiExtensionClient, ctx.clientPool
+
+	ns := createNamespaceOrDie("crd-cascading", clientSet, t)
+
+	definition, resourceClient := createRandomCustomResourceDefinition(t, apiExtensionClient, clientPool, ns.Name)
+
+	// Create a custom owner resource.
+	owner := newCRDInstance(definition, ns.Name, names.SimpleNameGenerator.GenerateName("owner"))
+	owner, err := resourceClient.Create(owner)
+	if err != nil {
+		t.Fatalf("failed to create owner resource %q: %v", owner.GetName(), err)
+	}
+	t.Logf("created owner resource %q", owner.GetName())
+
+	// Create a custom dependent resource.
+	dependent := newCRDInstance(definition, ns.Name, names.SimpleNameGenerator.GenerateName("dependent"))
+	link(t, owner, dependent)
+
+	dependent, err = resourceClient.Create(dependent)
+	if err != nil {
+		t.Fatalf("failed to create dependent resource %q: %v", dependent.GetName(), err)
+	}
+	t.Logf("created dependent resource %q", dependent.GetName())
+
+	// Delete the owner.
+	foreground := metav1.DeletePropagationForeground
+	err = resourceClient.Delete(owner.GetName(), &metav1.DeleteOptions{PropagationPolicy: &foreground})
+	if err != nil {
+		t.Fatalf("failed to delete owner resource %q: %v", owner.GetName(), err)
+	}
+
+	// Ensure the dependent is deleted.
+	if err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
+		_, err := resourceClient.Get(dependent.GetName(), metav1.GetOptions{})
+		return errors.IsNotFound(err), nil
+	}); err != nil {
+		t.Logf("owner: %#v", owner)
+		t.Logf("dependent: %#v", dependent)
+		t.Fatalf("failed waiting for dependent resource %q to be deleted", dependent.GetName())
+	}
+
+	// Ensure the owner is deleted.
+	if err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
```

Comment:
Since the policy=foreground, owner is expected to be deleted after the dependent is deleted. We should poll for owner's deletion first, then just check once that the dependent is deleted.